src/coreclr/tools/Common/Compiler/Dataflow/DynamicallyAccessedMembersBinder.cs

```diff
@@ -34,13 +34,15 @@ public static IEnumerable<TypeSystemEntity> GetDynamicallyAccessedMembers(this T
 
             if (memberTypes.HasFlag(DynamicallyAccessedMemberTypes.NonPublicConstructors))
             {
-                foreach (var c in typeDefinition.GetConstructorsOnType(filter: null, bindingFlags: BindingFlags.NonPublic))
+                bool withInherited = !declaredOnly && memberTypes.HasFlag(DynamicallyAccessedMemberTypesEx.NonPublicConstructorsWithInherited);
+                foreach (var c in typeDefinition.ApplyIncludeInherited(t => t.GetConstructorsOnType(filter: null, bindingFlags: BindingFlags.NonPublic), withInherited))
```

Comment:
I spent a good part of the day trying to figure out how to do this :)

The problem is that all of these APIs take BindingFlags, which is a public enum with documented meaning of all the flags. So:

* Private means: do not recurse in bases.
* DeclaredOnly also has a meaning for publics (it means do not recurse for publics)

I had an extra bool that says "please recurse into bases", but the extra argument then conflicts with the default reflection semantics and with whatever BindingFlags specify. This didn't feel particularly readable.

Then I considered just not using BindingFlags here, but we do read random BindingFlags using dataflow and just send them over to here so we need some translation to exist. We also sometimes inject DeclaredOnly by ourselves.

All in all what I have here felt like the cleanest because `WithInherited` behavior handling can be restricted to this one method instead of complicating the surface area everywhere.